File: hw/nase/u3/hw21AnimalFarm/Python/model/produce_animals.py
from datetime import date
from model.farm_animal import FarmAnimal
from model.interfaces import IProduceAnimal
from model.infrastructure import Product

class Chicken(FarmAnimal, IProduceAnimal):
    def __init__(self, id_animal: int, breed: str, born_on: date, weight: float, is_molting: bool, eggs_per_week: int):
        super().__init__(id_animal, breed, born_on, weight)
        self._is_molting = is_molting
        self._number_of_eggs_per_week = eggs_per_week

    def lay_an_egg(self) -> str:
        if not self._is_molting:
            return "The hen has successfully laid an egg.."
        return "The hen is molting; she cannot lay eggs at the moment.."

    def procedure(self) -> str:
        return "Manual egg collection in the poultry house."

    def measure_quantity(self, unit: str, quantity: float) -> None:
        self.product = Product(101, "Free-range eggs", unit, quantity)


class Sheep(FarmAnimal, IProduceAnimal):
    def __init__(self, id_animal: int, breed: str, born_on: date, weight: float, last_sheering: date):
        super().__init__(id_animal, breed, born_on, weight)
        self._last_sheering = last_sheering

    def procedure(self) -> str:
        return "Esquilado mecánico de lana fina."

    def measure_quantity(self, unit: str, quantity: float) -> None:
        self.product = Product(102, "Sheep's Wool", unit, quantity)

File: hw/nase/u3/hw21AnimalFarm/JavaScript/model/produceAnimals.js
const FarmAnimal = require('./farmAnimal');
const { Product } = require('./infrastructure');

class Chicken extends FarmAnimal {
    constructor(id, breed, bornOn, weight, isMolting, numberOfEggsPerWeek) {
        super(id, breed, bornOn, weight);
        this._isMolting = isMolting;
        this._numberOfEggsPerWeek = numberOfEggsPerWeek;
    }

    layAnEgg() {
        return !this._isMolting 
            ? "The hen has successfully laid an egg.." 
            : "The hen is molting; she cannot lay eggs at the moment..";
    }

    procedure() {
        return "Manual egg collection in the poultry house.";
    }

    measureQuantity(unit, quantity) {
        this.product = new Product(101, "Free-range eggs", unit, quantity);
    }
}

class Sheep extends FarmAnimal {
    constructor(id, breed, bornOn, weight, lastSheering) {
        super(id, breed, bornOn, weight);
        this._lastSheering = new Date(lastSheering);
    }

    procedure() {
        return "Mechanical shearing of fine wool.";
    }

    measureQuantity(unit, quantity) {
        this.product = new Product(102, "Lana de Oveja", unit, quantity);
    }
}

module.exports = { Chicken, Sheep };